README
Q Light Controller 3

Copyright (c) Heikki Junnila

EMail: hjunnila@users.sf.net
QLC at SourceForge: http://qlc.sf.net



License
-------
This program is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License
Version 2 as published by the Free Software Foundation.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details. The license is
in the file "COPYING".

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.


Requirements
------------
* QT 4.4 or newer


Compiling & Installation
------------------------

For windows, just install the latest QT package from Trolltech and
start the QT command prompt. For Linux or OSX, in addition to the QT
libraries, you need also qt development package.

See main/docs/compiling.html for more detailed information.
In general, just say:

	qmake

and then:

	make

as a regular user in the qlc root directory and then finally:

	make install

as root to install QLC.


Support & Bug Reports
---------------------
In general, support requests and bug reports should be sent to 
qlc-devel@lists.sf.net.

For TODO-lists, known bugs and other related information, go to
http://www.sf.net/projects/qlc


List of people who have contributed to this piece of software
-------------------------------------------------------------
Heikki Junnila (Author) <hjunnila@users.sf.net>
Stefan Krumm <stefankrumm@users.sf.net>
Christian Suehs <dance-or-die@users.sf.net>
Simon Newton <newtons@iinet.net>


ChangeLog
2008-10-10  Heikki Junnila  <hjunnila@users.sf.net>

	* main/addfixture.cpp:
	* main/app.cpp:
	* main/bus.cpp:
	* main/chaser.cpp:
	* main/chasereditor.cpp:
	* main/collection.cpp:
	* main/consolechannel.cpp:
	* main/doc.cpp:
	* main/efx.cpp:
	* main/efxfixture.cpp:
	* main/fixture.cpp:
	* main/fixtureconsole.cpp:
	* main/fixturemanager.cpp:
	* main/function.cpp:
	* main/functionmanager.cpp:
	* main/inputmanager.cpp:
	* main/inputmap.cpp:
	* main/keybind.cpp:
	* main/main.cpp:
	* main/monitor.cpp:
	* main/scene.cpp:
	* main/sceneeditor.cpp:
	* main/vcbutton.cpp:
	* main/vccuelist.cpp:
	* main/vcdockarea.cpp:
	* main/vcdockslider.cpp:
	* main/vcframe.cpp:
	* main/vclabel.cpp:
	* main/vcslider.cpp:
	* main/vcwidget.cpp:
	* main/vcxypad.cpp:
	* main/virtualconsole.cpp:

	Replaced cout with qDebug().
	Removed some whitespace.

2008-10-07  Heikki Junnila  <hjunnila@users.sf.net>

	* main/main.cpp
	* main/virtualconsole.cpp
	
	Removed key repeat code from apple builds.
	
2008-10-06  Heikki Junnila  <hjunnila@users.sf.net>

	* main/inputmanager.cpp
	* main/inputmanager.h
	* main/inputmanager.ui
	* main/inputpatcheditor.cpp
	* main/inputpatcheditor.h
	* main/inputpatcheditor.ui
	
	Removed the configuration button from input patch editor.
	Added "template" column to input manager.

2008-10-05  Heikki Junnila  <hjunnila@users.sf.net>

	* main/app.cpp

	Fixed application background restoration during startup.
	
2008-10-05  Heikki Junnila  <hjunnila@users.sf.net>

	* main/app.cpp
	* main/app.h
	* main/main.pro
	* main/pluginmanager.cpp
	* main/pluginmanager.h

	Removed plugin manager.
	Added input/output manager closing when going to operate mode.
	
2008-10-05  Heikki Junnila  <hjunnila@users.sf.net>

	* libs/common/qlcinplugin.h
	* libs/hidinput/hidinput.cpp
	* libs/hidinput/hidinput.h
	* libs/libs.pro
	* libs/midiinput/unix-midiinput.cpp
	* libs/midiinput/unix-midiinput.h
	* libs/midiinput/win32-midiinput.cpp
	* libs/midiinput/win32-midiinput.h
	* main/inputmanager.cpp
	* main/inputmap.cpp
	* main/inputmap.h
	* main/inputpatcheditor.cpp
	* main/inputpatcheditor.h
	* main/inputpatcheditor.ui

	Implemented the same kind of status info scheme to input plugins as
	output plugins.
	
2008-10-04  Heikki Junnila  <hjunnila@users.sf.net>

	* libs/llaout/llaout.cpp:
	* libs/llaout/llaout.h:
	* libs/usbdmxout/configureusbdmxout.cpp:
	* libs/usbdmxout/configureusbdmxout.h:
	* libs/usbdmxout/configureusbdmxout.ui:
	* libs/usbdmxout/usbdmxdevice-unix.cpp:
	* libs/usbdmxout/usbdmxdevice-unix.h:
	* libs/usbdmxout/usbdmxout-unix.cpp:
	* libs/usbdmxout/usbdmxout-unix.h:
	* main/dummyoutplugin.cpp:
	* main/inputmap.cpp:
	* main/outputmanager.cpp:
	* main/outputmanager.ui:
	* main/outputpatcheditor.cpp:
	* main/outputpatcheditor.h:
	* main/outputpatcheditor.ui:

	Fixed things to compile and work on unix.
	Implemented output patch editor's configure button.
	
2008-10-04  Heikki Junnila  <hjunnila@users.sf.net>

	* libs/dmx4linuxout/dmx4linuxout.cpp
	* libs/libs.pro
	* libs/llaout/llaout.cpp
	* libs/llaout/llaout.h
	* libs/usbdmxout/usbdmxdevice-win32.cpp
	* libs/usbdmxout/usbdmxout-unix.cpp
	* libs/usbdmxout/usbdmxout-win32.cpp
	* main/dummyoutplugin.cpp
	* main/outputmap.cpp
	* main/outputmap.h
	* main/outputpatcheditor.cpp
	* main/outputpatcheditor.ui

	Added an information text viewer to OutputPatchEditor and made some
	changes to output plugins to display more specific info.
	
2008-10-04  Heikki Junnila  <hjunnila@users.sf.net>

	* main/outputmanager.cpp

	Changed output manager to display the output name instead of its ID.

2008-10-04  Heikki Junnila  <hjunnila@users.sf.net>

	* libs/common/qlcoutplugin.h
	* libs/dmx4linuxout/configuredmx4linuxout.cpp
	* libs/dmx4linuxout/dmx4linuxout.cpp
	* libs/dmx4linuxout/dmx4linuxout.h
	* libs/dmx4linuxout/dmx4linuxout.pro
	* libs/libs.pro
	* libs/llaout/configurellaout.cpp
	* libs/llaout/configurellaout.h
	* libs/llaout/configurellaout.ui
	* libs/llaout/llaout.cpp
	* libs/llaout/llaout.h
	* libs/llaout/llaout.pro
	* libs/usbdmxout/configureusbdmxout.cpp
	* libs/usbdmxout/usbdmxout-win32.cpp
	* libs/usbdmxout/usbdmxout-win32.h
	* main/dummyoutplugin.cpp
	* main/dummyoutplugin.h
	* main/outputmap.cpp

	Removed int return type from output plugin's read&write methods.
	Added output parameter to output plugin's read&write methods.
	
2008-10-04  Heikki Junnila  <hjunnila@users.sf.net>

	* libs/usbdmxout/usbdmxdevice-win32.cpp
	* libs/usbdmxout/usbdmxdevice-win32.h
	* libs/usbdmxout/usbdmxout-win32.cpp

	Implemented read/writeRange methods.
	
2008-10-04  Heikki Junnila  <hjunnila@users.sf.net>

	* libs/usbdmxout/configureusbdmxout.cpp
	* libs/usbdmxout/usbdmxdevice-win32.cpp
	* libs/usbdmxout/usbdmxdevice-win32.h
	* libs/usbdmxout/usbdmxout-win32.cpp
	* libs/usbdmxout/usbdmxout-win32.h

	Fixed usbdmxout to work on windows with the modified output plugin arch
	
2008-10-02  Heikki Junnila  <hjunnila@users.sf.net>

	* libs/common/qlcoutplugin.h:
	* libs/dmx4linuxout/dmx4linuxout.cpp:
	* libs/dmx4linuxout/dmx4linuxout.h:
	* libs/llaout/llaout.cpp:
	* libs/llaout/llaout.h:
	* libs/usbdmxout/usbdmxout-unix.cpp:
	* libs/usbdmxout/usbdmxout-unix.h:
	* libs/usbdmxout/usbdmxout-win32.cpp:
	* libs/usbdmxout/usbdmxout-win32.h:
	* main/dummyoutplugin.cpp:
	* main/dummyoutplugin.h:
	* main/outputmanager.cpp:
	* main/outputmap.h:
	* main/outputpatcheditor.cpp:
	* main/outputpatcheditor.h:

	Added output parameter to QLCOutPlugin::infoText().

2008-10-01  Heikki Junnila  <hjunnila@users.sf.net>

	* libs/dmx4linuxout/dmx4linuxout.cpp:
	* libs/dmx4linuxout/dmx4linuxout.h:
	* libs/llaout/llaout.cpp:
	* libs/llaout/llaout.h:
	* libs/usbdmxout/configureusbdmxout.cpp:
	* libs/usbdmxout/usbdmxdevice-unix.cpp:
	* libs/usbdmxout/usbdmxdevice-unix.h:
	* libs/usbdmxout/usbdmxout-unix.cpp:
	* libs/usbdmxout/usbdmxout-unix.h:
	* libs/usbdmxout/usbdmxout.cpp:
	* libs/usbdmxout/usbdmxout.h:
	* libs/usbdmxout/usbdmxout.pro:

	Fixed plugins to compile in unix target.


main/app.h
/*
  Q Light Controller
  app.h

  Copyright (c) Heikki Junnila

  This program is free software; you can redistribute it and/or
  modify it under the terms of the GNU General Public License
  Version 2 as published by the Free Software Foundation.

  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details. The license is
  in the file "COPYING".

  You should have received a copy of the GNU General Public License
  along with this program; if not, write to the Free Software
  Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
*/

#ifndef APP_H
#define APP_H

#include <QMainWindow>
#include <QString>
#include <QList>

#include "common/qlcinplugin.h"
#include "common/qlcoutplugin.h"

class QMessageBox;
class QToolButton;
class QFileDialog;
class QStatusBar;
class QMenuBar;
class QToolBar;
class QPixmap;
class QAction;
class QLabel;
class QTimer;
class QMenu;

class FunctionConsumer;
class FunctionManager;
class FixtureManager;
class VirtualConsole;
class DummyOutPlugin;
class DummyInPlugin;
class QLCDocBrowser;
class BusProperties;
class QLCFixtureDef;
class OutputManager;
class InputManager;
class QLCInPlugin;
class QLCPlugin;
class OutputMap;
class InputMap;
class Monitor;
class Doc;
class App;

const QString KApplicationNameLong  = "Q Light Controller";
const QString KApplicationNameShort = "QLC";
const QString KApplicationVersion   = QString("Version ") + QString(VERSION);
const int KApplicationDefaultWidth  = 800;
const int KApplicationDefaultHeight = 600;

class App : public QMainWindow
{
	Q_OBJECT

	/*********************************************************************
	 * Initialization
	 *********************************************************************/
public:
	App();
	~App();

protected:
	void init();
	void closeEvent(QCloseEvent*);

	/*********************************************************************
	 * Output mapping
	 *********************************************************************/
public:
	OutputManager* outputManager() const { return m_outputManager; }
	OutputMap* outputMap() const { return m_outputMap; }

protected:
	void initOutputMap();

protected slots:
	void slotOutputMapBlackoutChanged(bool state);
	void slotFlashBlackoutIndicator();
	void slotOutputManager();
	void slotOutputManagerDestroyed(QObject* outputManager);

protected:
	OutputManager* m_outputManager;
	OutputMap* m_outputMap;

	/*********************************************************************
	 * Input mapping
	 *********************************************************************/
public:
	InputManager* inputManager() const { return m_inputManager; }
	InputMap* inputMap() { return m_inputMap; }

protected:
	void initInputMap();

protected slots:
	void slotInputManager();
	void slotInputManagerDestroyed(QObject* inputManager);

protected:
	InputManager* m_inputManager;
	InputMap* m_inputMap;

	/*********************************************************************
	 * Function Consumer
	 *********************************************************************/
public:
	/** Get a pointer to the function runner object */
	FunctionConsumer* functionConsumer() { return m_functionConsumer; }

protected:
	/** Initialize the function runner object */
	void initFunctionConsumer();

protected:
	/** The function runner object */
	FunctionConsumer* m_functionConsumer;

	/*********************************************************************
	 * Doc
	 *********************************************************************/
public:
	Doc* doc() { return m_doc; }
	void newDocument();

protected slots:
	void slotDocModified(bool state);

protected:
	void initDoc();

protected:
	Doc* m_doc;
	
	/*********************************************************************
	 * Fixture definitions
	 *********************************************************************/
public:
	/** Load all fixture definitions */
	bool initFixtureDefinitions();

	/** Get a fixture definition by its manufacturer & model */
	QLCFixtureDef* fixtureDef(const QString& manufacturer,
				  const QString& model);

	/** Get a list of fixture definitions */
	QList <QLCFixtureDef*> *fixtureDefList() { return &m_fixtureDefList; }

protected:
	/** List of fixture definitions */
	QList <QLCFixtureDef*> m_fixtureDefList;

	/*********************************************************************
	 * Mode: operate or design
	 *********************************************************************/
public:
	enum Mode { Operate, Design };
	Mode mode() { return m_mode; }

signals:
	void modeChanged(App::Mode mode);

public slots:
	void slotModeOperate();
	void slotModeDesign();

protected:
	/** Main operating mode */
	Mode m_mode;

	/*********************************************************************
	 * Fixture Manager
	 *********************************************************************/
public:
	FixtureManager* fixtureManager() const { return m_fixtureManager; }

protected:
	FixtureManager* m_fixtureManager;
	
	/*********************************************************************
	 * Function Manager
	 *********************************************************************/
public:
	FunctionManager* functionManager() const { return m_functionManager; }

protected:
	FunctionManager* m_functionManager;

	/*********************************************************************
	 * Bus Manager
	 *********************************************************************/
public:
	BusProperties* busManager() const { return m_busManager; }

protected:
	BusProperties* m_busManager;

	/*********************************************************************
	 * Virtual Console
	 *********************************************************************/
public:
	VirtualConsole* virtualConsole() const { return m_virtualConsole; }

protected:
	void initVirtualConsole();

protected:
	VirtualConsole* m_virtualConsole;

	/*********************************************************************
	 * Monitor
	 *********************************************************************/
public:
	Monitor* monitor() const { return m_monitor; }

protected:
	Monitor* m_monitor;

	/*********************************************************************
	 * Help browser
	 *********************************************************************/	
public:
	QLCDocBrowser* docBrowser() const { return m_docBrowser; }

protected:
	QLCDocBrowser* m_docBrowser;

	/*********************************************************************
	 * Status bar
	 *********************************************************************/
protected:
	/** Flashing blackout indicator on the status bar */
	QLabel* m_blackoutIndicator;

	/** Periodic timer object for the flashing indicator */
	QTimer* m_blackoutIndicatorTimer;

	/** Mode indicator on the status bar */
	QLabel* m_modeIndicator;

	/*********************************************************************
	 * Menus & toolbars
	 *********************************************************************/	
protected:
	void initActions();
	void initMenuBar();
	void initToolBar();
	void initStatusBar();
	QMenuBar *menuBar();

public slots:
	bool slotFileNew();
	void slotFileOpen();
	void slotFileSave();
	void slotFileSaveAs();
	void slotFileDirectories();
	void slotFileQuit();

	void slotFixtureManager();
	void slotFixtureManagerDestroyed(QObject* object);
	void slotFunctionManager();
	void slotFunctionManagerDestroyed(QObject* object);
	void slotBusManager();
	void slotBusManagerDestroyed(QObject* object);

	void slotControlVirtualConsole();
	void slotVirtualConsoleClosed();
	void slotControlMonitor();
	void slotMonitorDestroyed(QObject* object);
	void slotControlBlackout();
	void slotControlPanic();

	void slotHelpIndex();
	void slotDocBrowserDestroyed(QObject* object);	
	void slotHelpAbout();
	void slotHelpAboutQt();

	void slotCustomContextMenuRequested(const QPoint&);

protected:
	QAction* m_fileNewAction;
	QAction* m_fileOpenAction;
	QAction* m_fileSaveAction;
	QAction* m_fileSaveAsAction;
	QAction* m_fileDirectoriesAction;
	QAction* m_fileSaveDefaultsAction;
	QAction* m_fileQuitAction;

	QAction* m_fixtureManagerAction;
	QAction* m_functionManagerAction;
	QAction* m_busManagerAction;
	QAction* m_inputManagerAction;
	QAction* m_outputManagerAction;

	QAction* m_modeOperateAction;
	QAction* m_modeDesignAction;

	QAction* m_controlVCAction;
	QAction* m_controlMonitorAction;
	QAction* m_controlBlackoutAction;
	QAction* m_controlPanicAction;

	QAction* m_helpIndexAction;
	QAction* m_helpAboutAction;
	QAction* m_helpAboutQtAction;

protected:
	QMenu* m_fileMenu;
	QMenu* m_managerMenu;
	QMenu* m_controlMenu;
	QMenu* m_modeMenu;
	QMenu* m_helpMenu;
	
	QToolBar* m_toolbar;

	/*********************************************************************
	 * Workspace background
	 *********************************************************************/	
public:
	void setBackgroundImage(QString path);

public slots:
	void slotSetBackgroundImage();
	void slotClearBackgroundImage();

protected:
	QString m_backgroundImage;
};

#endif

